### PR TITLE
Add support for merging user permissions with team permissions

### PR DESCRIPTION
**Pull request by mknowlton89**
### Features and Changes

This PR introduces updated logic to the `getUserPermissions` method in `back-end/src/util/organization.util.ts` file. This method is used in a few places in the application to build a user's permissions.

Previously, a user could have a global role, and then multiple project roles, and when evaluating whether a user has permission to do a certain item, if the action is taking place in a project, and the user has a project-level role, we'd use the permissions associated with that role to determine if the user could do the thing. If they didn't have a project-level role, we'd then use their global role to determine if the user could do the thing.

With teams, the user's permissions and ultimately answering the question of "can a user do the thing they're asking to do" get's a lot more challenging.

It gets challenging because now a user can have a global role, and multiple project-level roles, and then they can be added to an infinite number of `teams`, and each `team` can also have a global role, and project-level roles. 

### Simple Example
Ross is a member of an organization, and his global role is "collaborator", with no project-specific roles. 

Ross has also been added to the `engineering` team within GrowthBook, that has a global role of `engineer`.

In this case, when building Ross's permissions, we first build his global role, and then, we look to see what team(s) Ross is on, and we then look to see if any of the permissions given to Ross by his `engineer` role aren't present in his existing permissions, and if so, we add them. 

In this example, we're essentially overriding Ross's `collaborator` role, with an `engineering` role. This is clean because the `engineer` role builds upon the `collaborator` role.

### A slightly less-simple example
Where things get slightly more complicated is when a user's global role, and a team role don't build upon each other.

For instance, let's look at Ross's colleague, Chandler - Chandler's global role is "engineer", but late last year, Chandler did some Data Science courses, and is now helping out with analyzing experiment results.

As such, he was added to the `analysts` team which has a global role of `analyst`.

Now, when building Chandler's permissions, we can't just override his `engineer` permissions with his `analyst` permissions, because then he wouldn't be able to add/edit feature flags, which is his main job.

Instead, we build his permissions based on his `engineer` role, and then see what permissions he is granted via his `analyst` role that he doesn't already have, and add those to his `userPermissions`.

### A complicated example
Another co-worker, we'll call her Monica, has some complicated permissions. She has a global role of `engineer`, but she's pretty new, so she can only edit Feature Flags in her personal dev environment called `monicas_dev_env` environment. However, just like Chandler, she got promoted last year, so she got added to the `ab_test_engineers` team, which also has a global role of `engineer`, but those engineers are able to make changes in the `staging` environment.

Now, when we're going to build Monica's permissions, we first build them based on her global engineering role, and limit her environments to just `monicas_dev_env`, but then we loop through the teams she's on, and we see that the `ab_test_engineers` team also gives her access to the `staging` environment, so we add `staging` to her existing list of environment limits, rather than just overriding it. 

Now, Monica can add/edit feature flags in both `monicas_dev_env` and `staging`.

### Another complicated example
Monica's colleague, Joey,  is very similar, but a little more experience. He's an engineer, and he can edit feature flags in his local environment `joeys_dev_env`. However, Joey has been at the company for awhile, so he can actually make changes in production. As a result, he was added to the `prod_engineers` group that has a global role of `engineer`, but this group doesn't have any environment limitations.

Now, when building Joey's permissions, just like everyone else, we build his base permissions via his global `engineer` role and add his environment limitations. Then, when looping through his teams, we see that they don't have any env limitations, so, we remove all environments from the `environments` array that limits what environments array he can edit it, giving him full access.

(As an note here, the opposite is also true, if his user-level global role doesn't limit his access by environment, but a team he's on does, we will honor the more permissive role, and not limit his access in any environment)

### A few highlights
So, based on the examples above, teams can give organizations easy way to control permissions for large groups of people. By creating an 'engineers' team, and adding users to it, you can then easily update ALL engineer's permissions in 1 place, rather than having to do it manually for each user.

With that said, the permissions granted by a team are MERGED with a user's permissions. They don't override it. So if a user has a more permissive global role than a team they're on, the team role doesn't override their user-level permissions.

This same logic applies to project-level permissions. Teams can also give project-level permissions, but again, those permissions are merged with any user's project-level permissions.

There is some funky logic with respect to `engineer` and `experimenter` roles - these are the only two roles where the `limitAccessByEnvironment` property really does anything, so in the code, these roles are treated as equals, when building the user's `environments` array.

### Dependencies

While not dependent, this work won't make any changes until this PR is merged in and the flag is flipped to enable it in production: LINK COMING SOON

### Testing

I wrote an extensive test file for this logic at permissions.test.ts that handles simple cases all the way up to corner/edge cases. Please take a look at that.

### CODE REVIEW

File: packages/back-end/src/util/organization.util.ts
Comment (msamper) [inline packages/back-end/src/util/organization.util.ts:112]: So if we grant access to more environments via a team then we would honor that over the more restrictive environments list set globally. That makes sense

File: packages/back-end/src/util/organization.util.ts
Comment (msamper) [inline packages/back-end/src/util/organization.util.ts:107]: Why do we only do the limit access by environment checks below if the role is `engineer` or `experimenter`? Is it because these two roles would be the only ones able to work with environments anyway? If so, I wonder if we can abstract that logic to somewhere else that is shared because I imagine if we add more roles that can work with environments we'd want to have a const or a function we can use to perform the check

File: packages/back-end/src/util/organization.util.ts
Comment (msamper) [inline packages/back-end/src/util/organization.util.ts:129]: Why are we working with the index of the roles here? Are we making an assumption that the ordering of the roles indicates which is more permissive? If so, we may want to do this in a different way to avoid making that assumption since it could introduce some bugs if we don't adhere to that rule and to make this a bit more readable

File: packages/back-end/src/util/organization.util.ts
Comment (msamper) [inline packages/back-end/src/util/organization.util.ts:146]: Should we early return this `UserPermissions` object if we don't find the `memberInfo` or throw an error?

File: packages/back-end/src/util/organization.util.ts
Comment (msamper) [inline packages/back-end/src/util/organization.util.ts:164]: Is this logic for admin roles in the code? Was looking for something like `memberInfo?.role === 'admin'` but don't see it

File: packages/back-end/src/util/organization.util.ts
Comment (msamper) [inline packages/back-end/src/util/organization.util.ts:189]: Consider doing this within `mergePermissions` 
Seems like it might make more sense for that function to handle all permission merging logic in case we ever call this somewhere else and you can just call it recursively if there are project roles

**Review by msamper — COMMENTED** [review]
Mainly reviewed the new `mergePermissions` function and the updates to `getUserPermissions` since they seemed like the biggest changes in this PR. Reviewing the tests separately since they're a bit long and to unblock. Overall the logic makes sense to me but left some questions in the comments. Might be worth doing some early returns with conditional logic to reduce some of the if/else nesting to improve the readability

File: packages/back-end/src/util/organization.util.ts
Comment (mknowlton89) [reply packages/back-end/src/util/organization.util.ts:112]: Yup - this way, if globally they get access to "staging", and then they're also on  a team , which gives them access to "dev", they'll then be able to have access to both "dev", and "staging".

File: packages/back-end/src/util/organization.util.ts
Comment (mknowlton89) [reply packages/back-end/src/util/organization.util.ts:102]: @msamper That is correct - only `engineer` and `experimenter` roles can be environment limited.

I did a bit of digging, and it looks like on the frontend we have a method `roleSupportsEnvLimit` - I've moved that to the shared folder, so we can now use that here (and continue using it on the frontend).

Good suggestion - I hated the pattern above. It reads much better now.

File: packages/back-end/src/util/organization.util.ts
Comment (mknowlton89) [reply packages/back-end/src/util/organization.util.ts:146]: Great suggestion - that cleans up the code quite a bit - I can get rid of so many `?`'s lol. I also added a test for this use case.

File: packages/back-end/src/util/organization.util.ts
Comment (mknowlton89) [reply packages/back-end/src/util/organization.util.ts]: I did have it in, but ultimately removed it accidentally left the comment in. 

After reviewing the logic - I felt like there might be some corner cases where we wouldn't want to skip building the team permissions, so I removed it.

I'm having trouble recalling the specifics right now, unfortunately.

For now, I've just removed the comment.

File: packages/back-end/src/util/organization.util.ts
Comment (mknowlton89) [reply packages/back-end/src/util/organization.util.ts:125]: You're reading it correctly. The logic is a bit twisty - all of the roles, do get more permissive, with the exception of the `engineer` and `analyst` roles - those don't fit that pattern.

I agree that this could introduce bugs, though, I'm not sure of an alternative route.

I think this pattern can also make future updates for "custom" roles a bit more challenging as well.

I'll do some brainstorming on this (probably tomorrow AM when my mind is more fresh).

File: packages/back-end/src/util/organization.util.ts
Comment (mknowlton89) [reply packages/back-end/src/util/organization.util.ts]: I abstracted this logic out 1 level with a new method `mergeUserWithTeamPermissions` that encapsulates this logic so we have a single method we can call and pass in the userPermissions, memberInfo, and org, and merge it all together.

File: packages/back-end/types/organization.d.ts
Comment (jdorn) [inline packages/back-end/types/organization.d.ts:46]: use `roleSupportsEnvLimit` instead

File: packages/shared/src/permissions/permissions.utils.ts
Comment (jdorn) [inline packages/shared/src/permissions/permissions.utils.ts:8]: Instead of hardcoding, check if the role has any `ENV_SCOPED_PERMISSIONS`

File: packages/back-end/src/util/organization.util.ts
Comment (jdorn) [inline packages/back-end/src/util/organization.util.ts:89]: use `newPermissions` here

File: packages/back-end/src/util/organization.util.ts
Comment (jdorn) [inline packages/back-end/src/util/organization.util.ts:99]: Make a clone instead of mutating

Comment (jdorn) [reply packages/back-end/src/util/organization.util.ts:99]: Why is `as` needed here?

File: packages/back-end/src/util/organization.util.ts
Comment (jdorn) [inline packages/back-end/src/util/organization.util.ts:104]: Is this still needed with the early return above?

File: packages/back-end/src/util/organization.util.ts
Comment (jdorn) [inline packages/back-end/src/util/organization.util.ts:82]: This is way I would structure it
```ts
combineRoles(items: UserPermissions[]): UserPermissions {
  let permissions = roleToPermissionMap(items[0]);

  for(let i=1; i<items.length; i++) {
    permissions = mergeUserPermissions(permissions, roleToPermissionsMap(items[i]));
  }

  return permissions;
}

mergeUserPermissions(userPermissions1: UserPermissions, userPermissions2: UserPermissions): UserPermissions {
  // Global, call `mergeRoles()`
  // Loop through projects, call `mergeRoles()`
}

mergeUserPermission(userPermission1: UserPermission, userPermission2: UserPermission): UserPermission {
  // Merge the permissions for the roles
}
```

File: packages/back-end/src/util/organization.util.ts
Comment (mknowlton89) [reply packages/back-end/src/util/organization.util.ts]: Without the `as` assertion, this is throwing the following error.

"Element implicitly has an 'any' type because expression of type 'string' can't be used to index type 'Partial<Record<Permission, boolean>>'.
  No index signature with a parameter of type 'string' was found on type 'Partial<Record<Permission, boolean>>'.ts(7053)"
  
  I've updated this slightly [here](https://github.com/growthbook/growthbook/pull/1640/commits/ce7c333a49e7324e14b42f8990eed2733fdc51ca) with a `isPermission` check to ensure that the `permission` string is as actual permission.
  
  It still isn't great. Going to take a break and come back to this as I've been trying to think of a cleaner solution for 30 minutes without success.

File: packages/back-end/src/util/organization.util.ts
Comment (jdorn) [inline packages/back-end/src/util/organization.util.ts:181]: So if a user only has a global role of `experimenter` and is added to a team that sets them as a collaborator in project A, the user will be restricted in project A to a collaborator, right? So being added to a team may actually give you fewer permissions than you had before. Not saying this is wrong, just trying to wrap my head around it.

Comment (mknowlton89) [reply packages/back-end/src/util/organization.util.ts:181]: @jdorn That is correct, which seems to be in alignment with what is in the app today - if a user has an `experimenter` role, but is on a team that gives them `collaborator` permissions, when we call `hasPermissions`, we check if there was a project passed in, and evaluate if they have permissions based on that project, otherwise, we fall back to the user's global role.

If we do want to change that logic, I think we would change it within `shared/permissions.utils.ts` rather than here. We would likely want to check the user's global permission, and if those were sufficient, then check the project.

Comment (jdorn) [reply packages/back-end/src/util/organization.util.ts:181]: Before merging user/team permissions, fill any missing project-scoped roles with the global role.

Role:
- Global: engineer
- Project A: readonly
- Project B: undefined

Becomes:
- Global: engineer
- Project A: readonly
- Project B: engineer

Comment (mknowlton89) [reply packages/back-end/src/util/organization.util.ts]: @jdorn Just pushed up [these changes](https://github.com/growthbook/growthbook/pull/1640/commits/ce8975e29f55617885f41788389b59b265a6cf66). 1 test is still failing (correctly), but I haven't been able to resolve the code issue in order to get the test to pass.

In these changes, I'm looping through all projects that the user has project-level permissions for, and if they're on a team that doesn't have permissions for that project, I'm setting the project-level permissions equal to the global role of the team.

I'm also looping through all of the projects the team has project-level permissions for, and if the user doesn't have project-level permissions for that project, I'm setting them equal to the user's global role.

However, in the case where the user doesn't have a project-level role for a project, but the team does, the user's global permissions are getting updated with that of the team's project-level permission.

Comment (mknowlton89) [reply packages/back-end/src/util/organization.util.ts:185]: Update - I think I got the bug resolved [here](https://github.com/growthbook/growthbook/pull/1640/commits/07e5090cb3a3705966249a5697baab781f3b56d2).

Updated method:

```ts
function mergeUserAndTeamPermissions(
  userPermissions: UserPermissions,
  teamPermissions: UserPermissions
) {
  userPermissions.global = mergeUserPermissionObj([
    userPermissions.global,
    teamPermissions.global,
  ]);

  // If a user is on a team that has project-level permissions,
  // but they're user-level permissions don't have project-level permissions,
  // set the user's project-level permissions equal to their global permissions.
  for (const project in teamPermissions.projects) {
    if (!userPermissions.projects[project]) {
      userPermissions.projects[project] = {
        limitAccessByEnvironment:
          userPermissions.global.limitAccessByEnvironment,
        environments: userPermissions.global.environments,
        permissions: userPermissions.global.permissions,
      };
    }
  }

  // If the user is on a team that doesn't have project-level permissions that the user has,
  // set the team's project-level permissions equal to the team's global role
  for (const project in userPermissions.projects) {
    if (!teamPermissions.projects[project]) {
      teamPermissions.projects[project] = {
        limitAccessByEnvironment:
          teamPermissions.global.limitAccessByEnvironment,
        environments: teamPermissions.global.environments,
        permissions: teamPermissions.global.permissions,
      };
    }
  }

  // Loop through each team project and merge it's permissions with the user's project permissions
  for (const project in teamPermissions.projects) {
    userPermissions.projects[project] = mergeUserPermissionObj([
      userPermissions.projects[project],
      teamPermissions.projects[project],
    ]);
  }

  // Loop through each user project and merge it's permissions with the team's project permissions
  for (const project in userPermissions.projects) {
    userPermissions.projects[project] = mergeUserPermissionObj([
      userPermissions.projects[project],
      teamPermissions.projects[project],
    ]);
  }
}

Comment (mknowlton89) [reply packages/back-end/src/util/organization.util.ts:185]: It feels like we could potentially merge user projects and team projects to reduce the number of loops needed.

Comment (jdorn) [reply packages/back-end/src/util/organization.util.ts]: This doesn't look quite right.
- User has `readonly` global role
- Team has `engineer` global role and `readonly` role in Project A

I would expect the result to have `readonly` permissions in Project A.

But, I believe looking at the code above, it would result in `engineer` permissions in Project A instead.

This is because you are merging global permissions before projects.  If you move the global merge after the project loops, it should fix it.

Comment (mknowlton89) [reply packages/back-end/src/util/organization.util.ts]: Ah - I see. I've updated the code and added a test case specific to this.

**Review by jdorn — APPROVED** [review]

**Pull request merged**
